<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>域 Node.js v0.10.18 文档 中文版</title>
  <link rel="stylesheet" href="/public/api_assets/style.css">
  <link rel="stylesheet" href="/public/api_assets/sh.css">
  <link rel="canonical" href="http://nodejs.org/api/domain.html">
</head>
<body class="alt apidoc" id="api-section-domain">

    <div id="intro" class="interior">
        <a href="/api/" title="返回首页">
            <img id="logo" src="http://nodejs.org/images/logo-light.png" alt="node.js">
        </a>
    </div>
    <div id="content" class="clearfix">
        <div id="column2" class="interior">
            <ul>
                <li><a href="/api/" class="docs current">文档首页</a></li>
                <li><a href="http://nodejs.org/api/">英文版文档</a></li>
            </ul>
            <div style="margin-top:44px; font-size:12px;">
              <a rel="license" href="http://creativecommons.org/licenses/by-nc/3.0/deed.zh"><img alt="知识共享许可协议" style="border-width:0" src="http://i.creativecommons.org/l/by-nc/3.0/88x31.png" /></a><br />本作品采用<a rel="license" href="http://creativecommons.org/licenses/by-nc/3.0/deed.zh">知识共享署名-非商业性使用 3.0 未本地化版本许可协议</a>进行许可。
            </div>
        </div>

        <div id="column1" class="interior">
          <header>
            <h1>Node.js v0.10.18 手册 &amp; 文档</h1>
            <div id="gtoc">
              <p>
                <a href="index.html" name="toc">索引</a> |
                <a href="all.html">在单一页面中浏览</a> |
                <a href="domain.json">JSON格式</a>
              </p>
            </div>
            <hr>
          </header>

          <div id="toc">
            <h2>目录</h2>
            <ul>
<li><a href="#domain_3558">域</a><ul>
<li><a href="#domain_3559">警告: 不要忽视错误!</a></li>
<li><a href="#domain_error_593">对Error(错误)对象的内容添加</a></li>
<li><a href="#domain_3560">隐式绑定</a></li>
<li><a href="#domain_3561">显式绑定</a></li>
<li><a href="#domain_domain_create_593">domain.create()</a></li>
<li><a href="#domain_domain_593">类: Domain</a><ul>
<li><a href="#domain_domain_run_fn_593">domain.run(fn)</a></li>
<li><a href="#domain_domain_members_593">domain.members</a></li>
<li><a href="#domain_domain_add_emitter_593">domain.add(emitter)</a></li>
<li><a href="#domain_domain_remove_emitter_593">domain.remove(emitter)</a></li>
<li><a href="#domain_domain_bind_callback_593">domain.bind(callback)</a><ul>
<li><a href="#domain_3562">例子</a></li>
</ul>
</li>
<li><a href="#domain_domain_intercept_callback_593">domain.intercept(callback)</a><ul>
<li><a href="#domain_3563">例子</a></li>
</ul>
</li>
<li><a href="#domain_domain_enter_593">domain.enter()</a></li>
<li><a href="#domain_domain_exit_593">domain.exit()</a></li>
<li><a href="#domain_domain_dispose_593">domain.dispose()</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>

          <div id="apicontent">
            <div class="translate-section" data-hash="02cca00571635b304013bb0f1929d602">

<h1>域<span><a href="#domain_3558" id="domain_3558">#</a></span></h1>
</div>

<div class="translate-section" data-hash="14ae3b8a1560651cf34fa2e5562e7f27">

<pre><code>稳定度: 2 - 不稳定</code></pre>
</div>

<div class="translate-section" data-hash="4518ebd75856bc13fa183ae778a0f8cf">

<p>Domains 提供了一种方式，即以一个单一的组的形式来处理多个不同的IO操作。如果任何一个注册到domain的事件触发器或回调触发了一个‘error’事件，或者抛出一个错误，那么domain对象将会被通知到。而不是直接让这个错误的上下文从`process.on（&apos;uncaughtException&apos;）&apos;处理程序中丢失掉，也不会致使程序因为这个错误伴随着错误码立即退出。

</p>
</div>

<div class="translate-section" data-hash="9f7e65c02ceb7847e9d1b089fd9045db">

<h2>警告: 不要忽视错误!<span><a href="#domain_3559" id="domain_3559">#</a></span></h2>
</div>

<!-- type=misc -->

<div class="translate-section" data-hash="0e9443611b77944922566c8d53e84ffb">

<p>Domain error处理程序不是一个在错误发生时，关闭你的进程的替代品

</p>
</div>

<div class="translate-section" data-hash="d2e33c9dd4cbc57096f58347b126fd5e">

<p>基于&apos;抛出（throw）&apos;在JavaScript中工作的方式，几乎从来没有任何方式能够在‘不泄露引用，不造成一些其他种类的未定义的脆弱状态’的前提下，安全的“从你离开的地方重新拾起（pick up where you left off）”，

</p>
</div>

<div class="translate-section" data-hash="48116c9b8f54166db432d18a336ea5ed">

<p>响应一个被抛出错误的最安全方式就是关闭进程。当然，在一个正常的Web服务器中，你可能会有很多活跃的连接。由于其他触发的错误你去突然关闭这些连接是不合理。

</p>
</div>

<div class="translate-section" data-hash="31e72d6174069a5067433645717c9973">

<p>更好的方法是发送错误响应给那个触发错误的请求，在保证其他人正常完成工作时，停止监听那个触发错误的人的新请求。

</p>
</div>

<div class="translate-section" data-hash="3bc705aac660bdeab33e0469f620025f">

<p>在这种方式中，<code>域</code>使用伴随着集群模块，由于主过程可以叉新工人时，一个工人发生了一个错误。节点程序规模的多
机，终止代理或服务注册可以注意一下失败，并做出相应的反应。

</p>
</div>

<div class="translate-section" data-hash="697acd8be310c559e6f31ebfef28a638">

<p>举例来说，以下就不是一个好想法：

</p>
</div>

<div class="translate-section" data-hash="3b5ceef23f6efb9bf6390967ac13d044">

<p>var d = require(&apos;domain&apos;).create();
d.on(&apos;error&apos;, function(er) {
  // 这个错误不会导致进程崩溃，但是情况会更糟糕！
  // 虽然我们阻止了进程突然重启动，但是我们已经发生了资源泄露
  // 这种事情的发生会让我们发疯。
  // 不如调用 process.on(&apos;uncaughtException&apos;)!
  console.log(&apos;error, but oh well&apos;, er.message);
});
d.run(function() {
  require(&apos;http&apos;).createServer(function(req, res) {
    handleRequest(req, res);
  }).listen(PORT);
});
</p>
<pre><code>
&lt;!-- endsection --&gt;

&lt;!-- section:c00fe3bdad4d3b86dcb006c3a8c55c76 --&gt;

通过对域的上下文的使用，以及将我们的程序分隔成多个工作进程的反射，我们可以做出更加恰当的反应和更加安全的处理。

&lt;!-- endsection --&gt;

&lt;!-- section:1749533625128d664ce57f327a3eb683 --&gt;

```javascript
// 好一些的做法！

&lt;!-- endsection --&gt;

&lt;!-- section:1002b470aa3001d212344187000af555 --&gt;

var cluster = require(&apos;cluster&apos;);
var PORT = +process.env.PORT || 1337;

&lt;!-- endsection --&gt;

&lt;!-- section:2f82bd01f23b2a29fbda6b4d88878ba5 --&gt;

if (cluster.isMaster) {
  // 在工作环境中，你可能会使用到不止一个工作分支
  // 而且可能不会把主干和分支放在同一个文件中
  //
  //你当然可以通过日志进行猜测，并且对你需要防止的DoS攻击等不良行为实施自定义的逻辑
  //
  // 看集群文件的选项
  //
  // 最重要的是主干非常小，增加了我们抵抗以外错误的可能性。

&lt;!-- endsection --&gt;

&lt;!-- section:cb647d887934b4f503df68aec3e5bb82 --&gt;

  cluster.fork();
  cluster.fork();

&lt;!-- endsection --&gt;

&lt;!-- section:73e4dad498a584605744c08e8889acd2 --&gt;

  cluster.on(&apos;disconnect&apos;, function(worker) {
    console.error(&apos;disconnect!&apos;);
    cluster.fork();
  });

&lt;!-- endsection --&gt;

&lt;!-- section:1eb9320833f0f4735c2a02437b831763 --&gt;

} else {
  // 工作进程
  //
  // 这是我们出错的地方

&lt;!-- endsection --&gt;

&lt;!-- section:4e5957024feac866ed9559ec321d9348 --&gt;

  var domain = require(&apos;domain&apos;);

&lt;!-- endsection --&gt;

&lt;!-- section:3eb2409b6e7d143f8eea5a1e547278b1 --&gt;

  //看集群文件对于使用工作进程处理请求的更多细节，它是如何工作的，它的警告等等。

&lt;!-- endsection --&gt;

&lt;!-- section:dd3bdd23463afb9f443cd297ec716fb5 --&gt;

  var server = require(&apos;http&apos;).createServer(function(req, res) {
    var d = domain.create();
    d.on(&apos;error&apos;, function(er) {
      console.error(&apos;error&apos;, er.stack);

&lt;!-- endsection --&gt;

&lt;!-- section:e614fa9e1583991a0d9b352d554182ba --&gt;

    // 因为req和res在这个域存在之前就被创建，
    // 所以我们需要显式添加它们。
    // 详见下面关于显式和隐式绑定的解释。
    d.add(req);
    d.add(res);

&lt;!-- endsection --&gt;

&lt;!-- section:ec427556232abea87deeffa7ae5a5830 --&gt;

    // 现在在域里面运行处理器函数。
    d.run(function() {
      handleRequest(req, res);
    });
  });
  server.listen(PORT);
}

&lt;!-- endsection --&gt;

&lt;!-- section:8792f2c8a87d64cd7a3893d4f03b9c89 --&gt;

    // 这个部分不是很重要。只是一个简单的路由例子。
    // 你会想把你的超级给力的应用逻辑放在这里。
    function handleRequest(req, res) {
      switch(req.url) {
        case &apos;/error&apos;:
          // 我们干了一些异步的东西，然后。。。
          setTimeout(function() {
            // 呃。。。
            flerb.bark();
          });
          break;
        default:
          res.end(&apos;ok&apos;);
      }
    }</code></pre>
</div>

<div class="translate-section" data-hash="2ba1d22539f91150d8cbee9acfa5c346">

<h2>对Error(错误)对象的内容添加<span><a href="#domain_error_593" id="domain_error_593">#</a></span></h2>
</div>

<!-- type=misc -->

<div class="translate-section" data-hash="9be72a5668ad058b5ad924d1e621a076">

<p>每一次一个Error对象被导向经过一个域，它会添加几个新的字段。

</p>
</div>

<div class="translate-section" data-hash="e323c97ee3c31022fc417710702133c1">

<ul>
<li><code>error.domain</code> 第一个处理这个错误的域。</li>
<li><code>error.domainEmitter</code> 用这个错误对象触发&apos;error&apos;事件的事件分发器。</li>
<li><code>error.domainBound</code> 回调函数，该回调函数被绑定到域，并且一个错误会作为第一参数传递给这个回调函数。</li>
<li><code>error.domainThrown</code> 一个布尔值表明这个错误是否被抛出，分发或者传递给一个绑定的回调函数。</li>
</ul>
</div>

<div class="translate-section" data-hash="1f1c06884feee02f47aa40d89c737d21">

<h2>隐式绑定<span><a href="#domain_3560" id="domain_3560">#</a></span></h2>
</div>

<!--type=misc-->

<div class="translate-section" data-hash="98204cb7669529ab02fae399ab30e0ba">

<p>如果多个域正在被使用，那么所有的<strong>新</strong>EventEmitter对象（包括Stream对象，请求，应答等等）会被隐式绑定到它们被创建时的有效域。

</p>
</div>

<div class="translate-section" data-hash="5dac1fd1e174e84f50ea01c009fe2b5f">

<p>而且，被传递到低层事件分发请求的回调函数（例如fs.open，或者其它接受回调函数的函数）会自动绑定到有效域。如果这些回调函数抛出错误，那么这个域会捕捉到这个错误。

</p>
</div>

<div class="translate-section" data-hash="ca0784d2572514b17bc521fb51af3bf5">

<p>为了防止内存的过度使用，Domain对象自己不会作为有效域的子对象被隐式添加到有效域。因为如果这样做的话，会很容易影响到请求和应答对象的正常垃圾回收。

</p>
</div>

<div class="translate-section" data-hash="4558bbd521aafe97106762e353ef47ed">

<p>如果你<em>想</em>在一个父Domain对象里嵌套子Domain对象，那么你需要显式地添加它们。

</p>
</div>

<div class="translate-section" data-hash="b3b67bdb5e1f2ab3bcf3c6464fd3d509">

<p>隐式绑定将被抛出的错误和<code>&apos;error&apos;</code>事件导向到Domain对象的<code>error</code>事件，但不会注册到Domain对象上的EventEmitter对象，所以<code>domain.dispose()</code>不会令EventEmitter对象停止运作。隐式绑定只关心被抛出的错误和 <code>&apos;error&apos;</code>事件。

</p>
</div>

<div class="translate-section" data-hash="aff225bd9e1f616da920940f466d3adc">

<h2>显式绑定<span><a href="#domain_3561" id="domain_3561">#</a></span></h2>
</div>

<!--type=misc-->

<div class="translate-section" data-hash="f03595da674aa2a3fff58efcbcfde955">

<p>有时，正在使用的域并不是某个事件分发器所应属的域。又或者，事件分发器在一个域内被创建，但是应该被绑定到另一个域。

</p>
</div>

<div class="translate-section" data-hash="68b518ea7f7da0e7973f96e8f1dc59c8">

<p>例如，对于一个HTTP服务器，可以有一个正在使用的域，但我们可能希望对每一个请求使用一个不同的域。

</p>
</div>

<div class="translate-section" data-hash="650459a148e19dbda0b9077972207243">

<p>这可以通过显示绑定来达到。

</p>
</div>

<div class="translate-section" data-hash="506c2c0c7f5b70af3df68c45c46f45a7">

<p>例如：

</p>
</div>

<div class="translate-section" data-hash="87c6c9012a9094cf7e37746bad59828f">

<pre><code>serverDomain.run(function() {
  // 服务器在serverDomain的作用域内被创建
  http.createServer(function(req, res) {
    // req和res同样在serverDomain的作用域内被创建
    // 但是，我们想对于每一个请求使用一个不一样的域。
    // 所以我们首先创建一个域，然后将req和res添加到这个域上。
    var reqd = domain.create();
    reqd.add(req);
    reqd.add(res);
    reqd.on(&apos;error&apos;, function(er) {
      console.error(&apos;Error&apos;, er, req.url);
      try {
        res.writeHead(500);
        res.end(&apos;Error occurred, sorry.&apos;);
      } catch (er) {
        console.error(&apos;Error sending 500&apos;, er, req.url);
      }
    });
  }).listen(1337);    
});
```</code></pre>
</div>

<div class="translate-section" data-hash="377ec28b7929bf46d98b3ee7de8da0f8">

<h2>domain.create()<span><a href="#domain_domain_create_593" id="domain_domain_create_593">#</a></span></h2>
</div>

<div class="translate-section" data-hash="fe5a36137aa2fef064450887f67719e2">

<ul>
<li>return: {Domain}</li>
</ul>
</div>

<div class="translate-section" data-hash="11251416cbdba2c005e2fdc23940c97c">

<p>返回一个新的Domain对象。

</p>
</div>

<div class="translate-section" data-hash="1ef2a8646e2b0c1453616384d2c2b02f">

<h2>类: Domain<span><a href="#domain_domain_593" id="domain_domain_593">#</a></span></h2>
</div>

<div class="translate-section" data-hash="b2db30eb1dc0a78c5269bf1f91f2ddc3">

<p>Domain类封装了将错误和没有被捕捉的异常导向到有效对象的功能。

</p>
</div>

<div class="translate-section" data-hash="6593a0d60283ffa945f933f0cba754bd">

<p>Domain是 <a href="events.html#events_class_events_eventemitter">EventEmitter</a>类的一个子类。监听它的<code>error</code>事件来处理它捕捉到的错误。

</p>
</div>

<div class="translate-section" data-hash="dd31240e5a7ba9da2c5294690f5c9799">

<h3>domain.run(fn)<span><a href="#domain_domain_run_fn_593" id="domain_domain_run_fn_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="b5ab910e8244a617af74aff6ada292b0">

<ul>
<li><code>fn</code> {Function}</li>
</ul>
</div>

<div class="translate-section" data-hash="359cb576d464c1284f6927d1dc5c4f07">

<p>在域的上下文里运行提供的函数，隐式地绑定所有该上下文里创建的事件分发器，计时器和低层请求。

</p>
</div>

<div class="translate-section" data-hash="be5fcc18fabc06e1e1657a1654b98379">

<p>这是使用一个域的最基本的方式。

</p>
</div>

<div class="translate-section" data-hash="81eeab9506186e2dca8faefa78d54067">

<p>实例：

</p>
</div>

<div class="translate-section" data-hash="67d1177693bf8b9111a2e239af88ee60">

<pre><code>var d = domain.create();
d.on(&apos;error&apos;, function(er) {
  console.error(&apos;Caught error!&apos;, er);
});
d.run(function() {
  process.nextTick(function() {
    setTimeout(function() { // 模拟几个不同的异步的东西
      fs.open(&apos;non-existent file&apos;, &apos;r&apos;, function(er, fd) {
        if (er) throw er;
        // 继续。。。
      });
    }, 100);
  });
});</code></pre>
</div>

<div class="translate-section" data-hash="6c237a25cfa7701779036f0391f523cf">

<p>在这个例子里， <code>d.on(&apos;error&apos;)</code> 处理器会被触发，而不是导致程序崩溃。

</p>
</div>

<div class="translate-section" data-hash="5e0a823fb76f4401513ad0621aeb1a99">

<h3>domain.members<span><a href="#domain_domain_members_593" id="domain_domain_members_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="da988593f493b305c0e8d1e200e041c4">

<ul>
<li>{Array}</li>
</ul>
</div>

<div class="translate-section" data-hash="686ace803edf248d64029d99facc6b1a">

<p>一个数组，里面的元素是被显式添加到域里的计时器和事件分发器。

</p>
</div>

<div class="translate-section" data-hash="9784a910a37bca98cae641f0083a4f02">

<h3>domain.add(emitter)<span><a href="#domain_domain_add_emitter_593" id="domain_domain_add_emitter_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="6bc3b39d7a33fef945ea0c098c460829">

<ul>
<li><code>emitter</code> {EventEmitter | Timer} 被添加到域里的时间分发器或计时器</li>
</ul>
</div>

<div class="translate-section" data-hash="7cdb670833b897428c6c913dbc2784f3">

<p>显式地将一个分发器添加到域。如果这个分发器调用的任意一个事件处理器抛出一个错误，或是这个分发器分发了一个<code>error</code>事，那么它会被导向到这个域的<code>error</code>事件，就像隐式绑定所做的一样。

</p>
</div>

<div class="translate-section" data-hash="2f79dc18affd0c5bf1452fa21ff190ba">

<p>这对于从<code>setInterval</code>和<code>setTimeout</code>返回的计时器同样适用。如果这些计时器的回调函数抛出错误，它将会被这个域的<code>error</code>处理器捕捉到。

</p>
</div>

<div class="translate-section" data-hash="8284eecba7860093ee169ecc6e94d239">

<p>如果这个Timer或EventEmitter对象已经被绑定到另外一个域，那么它将会从那个域被移除，然后绑定到当前的域。

</p>
</div>

<div class="translate-section" data-hash="9484273a4a36baa279f52f6d714fe716">

<h3>domain.remove(emitter)<span><a href="#domain_domain_remove_emitter_593" id="domain_domain_remove_emitter_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="4fcd7597529011cf55cc89c89f8f8a83">

<ul>
<li><code>emitter</code> {EventEmitter | Timer} 要从域里被移除的分发器或计时器</li>
</ul>
</div>

<div class="translate-section" data-hash="996999db3f818534856cd51565dee5f1">

<p>与<code>domain.add(emitter)</code>函数恰恰相反，这个函数将域处理从指明的分发器里移除。

</p>
</div>

<div class="translate-section" data-hash="2e6b4d734d494f2d5b37cf1bb6ab6cdd">

<h3>domain.bind(callback)<span><a href="#domain_domain_bind_callback_593" id="domain_domain_bind_callback_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="9f1fe1b00f5dcc5436d12215a72ff025">

<ul>
<li><code>callback</code> {Function} 回调函数</li>
<li>return: {Function} 被绑定的函数</li>
</ul>
</div>

<div class="translate-section" data-hash="e3c7108af2a21c27c006df64df0f1268">

<p>返回的函数会是一个对于所提供的回调函数的包装函数。当这个被返回的函数被调用时，所有被抛出的错误都会被导向到这个域的<code>error</code>事件。

</p>
</div>

<div class="translate-section" data-hash="0d7592e22a9740a33e30f517f476ebb9">

<h4>例子<span><a href="#domain_3562" id="domain_3562">#</a></span></h4>
</div>

<div class="translate-section" data-hash="82e5af0bbac777ea0f78400ba759a4fb">

<pre><code>d.on(&apos;error&apos;, function(er) {
  // 有个地方发生了一个错误。
  // 如果我们现在抛出这个错误，它会让整个程序崩溃
  // 并给出行号和栈信息。
});</code></pre>
</div>

<div class="translate-section" data-hash="d760cee1417aa31ed1a640282afe5537">

<h3>domain.intercept(callback)<span><a href="#domain_domain_intercept_callback_593" id="domain_domain_intercept_callback_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="15bc26c472072a403d639e5a9eacd48a">

<ul>
<li><code>callback</code> {Function} 回调函数</li>
<li>return: {Function} 被拦截的函数</li>
</ul>
</div>

<div class="translate-section" data-hash="150ee58532d7a6bdd44288a3579cccb7">

<p>这个函数与<code>domain.bind(callback)</code>几乎一模一样。但是，除了捕捉被抛出的错误外，它还会拦截作为第一参数被传递到这个函数的<code>Error</code>对象。

</p>
</div>

<div class="translate-section" data-hash="3de54f51f66d21f0150ed66612e30db5">

<p>在这种方式下，常见的&apos;if(er) return callback(er);&apos;的方式可以被一个单独地方的单独的错误处理所取代。

</p>
</div>

<div class="translate-section" data-hash="0d7592e22a9740a33e30f517f476ebb9">

<h4>例子<span><a href="#domain_3563" id="domain_3563">#</a></span></h4>
</div>

<div class="translate-section" data-hash="82e5af0bbac777ea0f78400ba759a4fb">

<pre><code>d.on(&apos;error&apos;, function(er) {
  // 有个地方发生了一个错误。
  // 如果我们现在抛出这个错误，它会让整个程序崩溃
  // 并给出行号和栈信息。
});</code></pre>
</div>

<div class="translate-section" data-hash="96e9f23d2963452708a111413212b7b0">

<h3>domain.enter()<span><a href="#domain_domain_enter_593" id="domain_domain_enter_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="939be1353d5589f7603f8a4322b2cdcd">

<p><code>enter</code>函数对于<code>run</code>，<code>bind</code>和<code>intercept</code>来说就像它们的管道系统：它们使用<code>enter</code>函数来设置有效域。<code>enter</code>函数对于域设定了<code>domain.active</code>和 <code>process.domain</code>
，还隐式地将域推入了由域模块管理的域栈（关于域栈的细节详见<code>domain.exit()</code>）。<code>enter</code>函数的调用，分隔了异步调用链以及绑定到一个域的I/O操作的结束或中断。

</p>
</div>

<div class="translate-section" data-hash="82a76791963a2eba5fb3e7850ad8379a">

<p>调用<code>enter</code>仅仅改变活动的域，而不改变域本身。
<code>Enter</code> 和 <code>exit</code>在一个单独的域可以被调用任意多次。

</p>
</div>

<div class="translate-section" data-hash="2b6570c84523cc71c54b20aeb4133e52">

<p>如果域的<code>enter</code>已经设置，<code>enter</code>将不设置域就返回。

</p>
</div>

<div class="translate-section" data-hash="b5dad78a6027862ce2987d5bd6f1fdc3">

<h3>domain.exit()<span><a href="#domain_domain_exit_593" id="domain_domain_exit_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="21a50c8a3268fd6ebb5479e8f71c799a">

<p><code>exit</code>函数退出当前的域，将当前域从域的栈里移除。每当当程序的执行流程准要切换到一个不同的异步调用链的上下文时，要保证退出当前的域。<code>exit</code>函数的调用，分隔了异步调用链以及绑定到一个域的I/O操作的结束或中断。


</p>
</div>

<div class="translate-section" data-hash="82e4ab2dffeb19245d9ace72bb755fee">

<p>如果有多个嵌套的域绑定到当前的执行上下文，
<code>退出</code>将退出在这个域里的所有的嵌套。

</p>
</div>

<div class="translate-section" data-hash="48c8839c66b3ed4b5ce0104a03aa4cec">

<p>调用<code>exit</code>只会改变有效域，而不会改变域自身。在一个单一域上，<code>Enter</code>和<code>exit</code>可以被调用任意次。

</p>
</div>

<div class="translate-section" data-hash="01f107e0a9db2ebe61f8f68bedca7b2a">

<p>如果在这个域名下<code>exit </code>已经被设置，<code>exit </code>将不退出域返回。

</p>
</div>

<div class="translate-section" data-hash="e6e28b09e96c1b7e9fea0ea8b0ba669d">

<h3>domain.dispose()<span><a href="#domain_domain_dispose_593" id="domain_domain_dispose_593">#</a></span></h3>
</div>

<div class="translate-section" data-hash="1685ae170cabba71590ec4d8e306d691">

<pre><code>稳定度: 0 - 已过时。请通过设置在域上的错误事件处理器，显式地东失败的IO操作中恢复。</code></pre>
</div>

<div class="translate-section" data-hash="40376032040e0f5516c09961eb69326f">

<p>一旦<code>dispose</code>被调用，通过<code>run</code>，<code>bind</code>或<code>intercept</code>绑定到这个域的回调函数将不再使用这个域，并且一个<code>dispose</code>事件会被分发。

</p>
</div>

<div class="translate-section" data-hash="6949f457689f76a715805ad76aa0f835">

</div>
          </div>
        </div>
    </div>
    <div id="footer">
        <a href="http://joyent.com" class="joyent-logo">Joyent</a>
        <ul class="clearfix">
            <li><a href="http://nodejs.org/">Node.js</a></li>
            <li><a href="http://nodejs.org/download/">Download</a></li>
            <li><a href="http://nodejs.org/about/">About</a></li>
            <li><a href="http://npmjs.org/">npm Registry</a></li>
            <li><a href="http://nodejs.org/api/">Docs</a></li>
            <li><a href="http://blog.nodejs.org">Blog</a></li>
            <li><a href="http://nodejs.org/community/">Community</a></li>
            <li><a href="http://nodejs.org/logos/">Logos</a></li>
            <li><a href="http://jobs.nodejs.org/">Jobs</a></li>
            <li><a href="http://twitter.com/nodejs" class="twitter">@nodejs</a></li>
        </ul>

        <p>Copyright <a href="http://joyent.com/">Joyent, Inc</a>, Node.js is a <a href="/trademark-policy.pdf">trademark</a> of Joyent, Inc. View <a href="https://raw.github.com/joyent/node/v0.10.18/LICENSE">license</a>.</p>
    </div>

  <script src="/public/api_assets/sh_main.js"></script>
  <script src="/public/api_assets/sh_javascript.min.js"></script>
  <script>highlight(undefined, undefined, 'pre');</script>
</body>
</html>

<!-- google analytics -->
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-26599868-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>

<script src="/public/js/jquery.js"></script>
<script>
$('.translate-section').each(function () {
  var html = '<div class="op">' +
               '<button class="show-origin">显示原文</button>' +
               '<button class="btn btn-xs btn-info">其他翻译</button>' +
               '<button class="btn btn-xs btn-info">纠错</button>' +
              '</div>';
  $(this).prepend(html);
});

// 显示原文
$('.translate-section').delegate('.show-origin', 'click', function () {
  var $section = $(this).closest('.translate-section');
  var hash = $section.data('hash');
  $.get('/translate/get/origin', {hash: hash}, function (d) {
    if (d.error) return alert(d.error);

    $section.find('.origin').remove();
    $section.append('<div class="origin">' + d.html + '</div>');
    highlight(undefined, undefined, 'code');
  });
});
</script>